### PR TITLE
[Typescript] Add application/x-www-form-urlencoded content support

### PR DESCRIPTION
**Pull request by ksvirkou-hubspot**
https://github.com/OpenAPITools/openapi-generator/issues/9962

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.3.x`, `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by ksvirkou-hubspot** [comment]
@TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx @macjohnny @topce @akehir @petejohansonxo @amakhrov

**Comment by ksvirkou-hubspot** [comment]
Any updates on this?
@TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx @macjohnny @topce @akehir @petejohansonxo @amakhrov

**Comment by atanasiuk-hubspot** [comment]
Hi, team! Any feedback on this?

**Comment by ksvirkou-hubspot** [comment]
Hi
Any feedback on this?
@TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx @macjohnny @topce @akehir @petejohansonxo @amakhrov

File: samples/openapi3/client/petstore/typescript/builds/default/util.ts [hunk 1]
```diff
@@ -26,3 +26,17 @@ export function isCodeInRange(codeRange: string, code: number): boolean {
...
+*
+* @param consumes array
+*/
+export function canConsumeForm(contentTypes: string[]): boolean {
+    for (const contentType of contentTypes) {
+        if ('multipart/form-data' === contentType) {
+            return true;
+        }
+    }
+    return false;
```
Comment (amakhrov) [inline samples/openapi3/client/petstore/typescript/builds/default/util.ts:41]: is it the same as just `return contentTypes.includes('multipart/form-data')`?

File: samples/openapi3/client/petstore/typescript/builds/default/apis/PetApi.ts [hunk 2]
```diff
@@ -364,12 +339,17 @@ export class PetApiRequestFactory extends BaseAPIRequestFactory {
...
+        const useForm = canConsumeForm([
+            'multipart/form-data',
+        ]);
+
+        let localVarFormParams
+        if (useForm) {
+            localVarFormParams = new FormData();
+        } else {
+            localVarFormParams = new URLSearchParams();
+        }
```
Comment (amakhrov) [inline samples/openapi3/client/petstore/typescript/builds/default/apis/PetApi.ts:352]: I'm wondering whether it would make more sense to handle the conditions during code generation? The content types are known from the spec, at generation time. It would make the emitted code simpler.

File: samples/server/petstore/haskell-servant/lib/OpenAPIPetstore/API.hs [hunk 3]
```diff
@@ -284,7 +286,13 @@ runOpenAPIPetstoreMiddlewareServer Config{..} middleware backend = do
...
         & Warp.setPort (baseUrlPort url)
         & Warp.setHost (fromString $ baseUrlHost url)
-  liftIO $ Warp.runSettings warpSettings $ middleware $ serve (Proxy :: Proxy OpenAPIPetstoreAPI) (serverFromBackend backend)
+  liftIO $ Warp.runSettings warpSettings $ middleware $ serverWaiApplicationOpenAPIPetstore backend
+
+-- | Plain "Network.Wai" Application for the OpenAPIPetstore server.
+--
+-- Can be used to implement e.g. tests that call the API without a full webserver.
+serverWaiApplicationOpenAPIPetstore :: OpenAPIPetstoreBackend (ExceptT ServerError IO) -> Application
+serverWaiApplicationOpenAPIPetstore backend = serve (Proxy :: Proxy OpenAPIPetstoreAPI) (serverFromBackend backend)
```
Comment (amakhrov) [inline samples/server/petstore/haskell-servant/lib/OpenAPIPetstore/API.hs:295]: this file looks unrelated - probably need to revert

**Review by amakhrov — COMMENTED** [review]
Thanks for your contribution!

Overall the PR looks good. I'm wondering if you considered defining the conditional code generation instead of having it at run time in the generated code

**Comment by macjohnny** [comment]
@ksvirkou-hubspot thanks for your contribution. after you have applied the changes suggested by @amakhrov, I am happy to merge.

File: samples/openapi3/client/petstore/typescript/builds/default/apis/PetApi.ts (hunk 2 above)
Comment (ksvirkou-hubspot) [reply samples/openapi3/client/petstore/typescript/builds/default/apis/PetApi.ts:352]: I've thought about it but
We did in a similar way as in [typescript-fetch](https://github.com/OpenAPITools/openapi-generator/pull/3934) and [typescript-angular](https://github.com/OpenAPITools/openapi-generator/blob/da5d611ca98a42f69d21fb844dfb4e26fd558e54/modules/openapi-generator/src/main/resources/typescript-angular/api.service.mustache#L84)
I can do it as you proposed 
But I don’t know what would be better 
what do you think? 
CC @amakhrov @macjohnny

File: samples/openapi3/client/petstore/typescript/builds/default/util.ts (hunk 1 above)
Comment (ksvirkou-hubspot) [reply samples/openapi3/client/petstore/typescript/builds/default/util.ts]: I try to use includes and I got an error 
`Property ‘includes’ does not exist on type ‘string[]’. Do you need to change your target library? Try changing the ‘lib’ compiler option to ‘es2016’ or later.ts`
it breaks some apps I think  
lets use the old version or `arr.indexOf(valueToCheck) !== -1`
What do you think?
CC @amakhrov @macjohnny

Comment (amakhrov) [reply samples/openapi3/client/petstore/typescript/builds/default/util.ts]: `indexOf` sounds good.

File: samples/openapi3/client/petstore/typescript/builds/default/apis/PetApi.ts (hunk 2 above)
Comment (amakhrov) [reply samples/openapi3/client/petstore/typescript/builds/default/apis/PetApi.ts:352]: I believe preferring simplicity in the end code over simplicity in the codegen template is the right tradeoff - hence handling the conditions during code generation is better.
But I see how it can increase the scope of PR here.
I totally don't object keeping it the way it is right now

Comment (ksvirkou-hubspot) [reply samples/openapi3/client/petstore/typescript/builds/default/apis/PetApi.ts:352]: let's leave it the way it is right now
because it will increase the scope of PR here.
CC @amakhrov @macjohnny

**Comment by ksvirkou-hubspot** [comment]
> @ksvirkou-hubspot thanks for your contribution. after you have applied the changes suggested by @amakhrov, I am happy to merge.
@macjohnny @amakhrov 
Could you merge it ?
it would be great to merge it

**Comment by macjohnny** [comment]
Somehow the samples are not up to date. Please regenerate them

**Comment by ksvirkou-hubspot** [comment]
> Somehow the samples are not up to date. Please regenerate them

It isn't my blame
I can fix it but it is unrelated with this pr
more over @amakhrov asked me that I had not done it (could you check the third comment of @amakhrov? )
CC @macjohnny

**Comment by macjohnny** [comment]
You can merge the most recent master into your branch, rebuild, re-generate the samples, and commit everything

**Comment by ksvirkou-hubspot** [comment]
> You can merge the most recent master into your branch, rebuild, re-generate the samples, and commit everything

Done 
CC @macjohnny

**Comment by wing328** [comment]
@ksvirkou-hubspot FYI. @macjohnny filed https://github.com/OpenAPITools/openapi-generator/pull/10288 to fix an issue introduced by this PR. Please test with the latest master and let us know if you've questions on the fix.

**Comment by ksvirkou-hubspot** [comment]
> @ksvirkou-hubspot FYI. @macjohnny filed #10288 to fix an issue introduced by this PR. Please test with the latest master and let us know if you've questions on the fix.

it works
I left a comment [here](https://github.com/OpenAPITools/openapi-generator/pull/10288)
